### ISSUE TITLE
Saving the form invalidates the signature of the PDF file when the PDF is a form

### ISSUE THREAD

**Issue by fgerges** [mozilla/pdf.js#14816]
Attach (recommended) or Link to PDF file here: see attachment

Configuration:
- Chrome 100.0.4896.127 64 bits and Firefox 100.0b8 64 bits
- Operating system : MS Windows 10 Pro 21H2 19044.1645
- PDF.js version 2.13 and 2.14 and tested also on https://mozilla.github.io/pdf.js/web/viewer.html

Steps to reproduce the problem:
1. set adobe acrobat reader as default viewer for PDF file to activate it on pdf file download
2. go on https://mozilla.github.io/pdf.js/web/viewer.html
3. open the pdf in attachment
4. enter information in the fields of the form
5. click on the download button (which saves the pdf and triggers its download and then its opening in Adobe acrobat reader)

What is the expected behavior? (done with Adobe Acrobat Reader 2022.001.20117 64 bits)
![image](https://user-images.githubusercontent.com/68105056/164455308-b07dae48-999f-4b4d-8121-da4ca5850fb6.png)

What went wrong?
![image](https://user-images.githubusercontent.com/68105056/164455600-2c7bd20f-6ed3-46a1-b12a-b419bfdddacb.png)

[formulaire_signé.pdf](https://github.com/mozilla/pdf.js/files/8530689/formulaire_signe.pdf)

**Comment by Snuffleupagus** [comment]
Duplicate of #13351

**Comment by marco-c** [comment]
Are you seeing a different behavior if you save with Adobe Reader? I would expect the signature to be stripped since you are effectively modifying the PDF

**Comment by maxkurze1** [comment]
> That being said, I don't think this is a good idea to fix this issue.
>
> This might be used to trick users into believing that a document with unsigned amendments constitutes its actually signed version

I disagree with your opinion.

#### the problem is the visualization

I can totally relate the problem of having a green confirmation on a document which contains unsigned content, but you are drawing the wrong conclusion here.

The thing that is misleading is the *visualization* itself, i.e. having a 'green confirmation' on something that is partially signed. It is not the state of the document that's in fault here.

If someone intended to trick people, they could very well generate that partially signed document using other tools. But what actually confuses people is when pdf.js just accepts such a document as 'valid'.

#### the signature chooses acceptable modifications

The signer (when creating the signature) is responsible to choose which additional modifications are acceptable and which should invalidate the signatures.

Thus, it is not even our duty to assess whether the signature should be display as valid or invalid. In case the signer chooses to accept form fillings after the signature's creation, then it is not misleading to show his signature as valid - it is simply representing his intention.

#### documents with multiple revisions

Certain documents are just intended to be signed multiple times, while form fields are typically filled in between. Just think of a contract (like a rental agreement) where multiple people have to enter their data and sign. The typical workflow would always go: person1 enters data -> person1 signs -> p2 enters data -> p2 signs -> ... .

In such a case, the form filling should definitely preserve the first signature's validity.

#### conclusion

That being said, in case a document already contains a signature, pdf.js should just respect it and use an incremental update for the form fillings instead of a normal one. For me, there is no imaginable way how that could do any harm, except that the PDFs file size is going to increase a little.

Additionally, I'd argue that this issue does not depend on #13351. It is not necessary for this enhancement to have signature validation is place. IMO, it should be enough to recognize that there _exists_ a signature (valid or invalid), in order to make an incremental update on the side of caution. Even if the signature would be deemed invalid by some pdf.js validation algorithm, we should keep its signed content unmodified in case someone else (or some other tool) actually intends it to be this way.
